core/src/main/java/org/apache/iceberg/io/ObjectModelFactory.java

```diff
@@ -0,0 +1,126 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io;
+
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.deletes.PositionDelete;
+
+/**
+ * Interface that provides a unified abstraction for converting between data file formats and
+ * input/output data representations.
+ *
+ * <p>{@link ObjectModelFactory} serves as a bridge between storage formats ({@link FileFormat}) and
+ * expected input/output data structures, optimizing performance through direct conversion without
+ * intermediate representations. File format implementations handle the low-level parsing details
+ * while the object model determines the in-memory representation used for the parsed data.
+ * Together, these provide a consistent API for consuming data files while optimizing for specific
+ * processing engines.
+ *
+ * <p>Iceberg provides some built-in object models and processing engines can implement custom
+ * object models to integrate with Iceberg's file reading and writing capabilities.
+ *
+ * @param <E> input schema type used when converting input data to the file format
+ * @param <D> output type used for reading data, and input type for writing data and deletes
+ */
+public interface ObjectModelFactory<E, D> {
+  /** The file format which is read/written by the object model. */
+  FileFormat format();
+
+  /**
+   * Returns the unique identifier for the object model implementation processed by this factory.
+   *
+   * <p>The object model names (for example: "generic", "spark", "spark-vectorized", "flink",
+   * "arrow") act as a contract specifying the expected data structures for both reading (converting
+   * file formats into output objects) and writing (converting input objects into file formats).
+   * This ensures proper integration between Iceberg's storage layer and processing engines.
+   *
+   * @return string identifier for this object model implementation
+   */
+  String objectModelName();
+
+  /**
+   * Creates a writer builder for standard data files.
+   *
+   * <p>The returned {@link WriteBuilder} configures and creates a writer that converts input
+   * objects into the file format supported by this factory for regular data content.
+   *
+   * <p>The builder follows the fluent pattern for configuring writer properties like compression,
+   * encryption, row group size, and other format-specific options.
+   *
+   * @param outputFile destination for the written data
+   * @return configured writer builder for standard data files
+   * @param <B> the concrete builder type for method chaining
+   */
+  <B extends WriteBuilder<B, E, D>> B dataWriteBuilder(OutputFile outputFile);
```

Comment:
Minor: I'd prefer shorter names here as well. The formats use `writeData` and `writeDeletes`, although that doesn't necessarily make sense here. What about `dataBuilder`, `deleteBuilder`, and `readBuilder`?

Reply:
Done:
- dataBuilder
- equalityDeleteBuilder
- positionDeleteBuilder
- readBuilder